tuf/repository/_repository.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2021-2022 python-tuf contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Repository Abstraction for metadata management"""
+
+import logging
+from abc import ABC, abstractmethod
+from contextlib import contextmanager, suppress
+from copy import deepcopy
+from typing import Dict, Generator, Optional, Tuple
+
+from tuf.api.metadata import Metadata, MetaFile, Signed
+
+logger = logging.getLogger(__name__)
+
+
+class AbortEdit(Exception):
+    """Raise to exit the edit() contextmanager without saving changes"""
+
+
+class Repository(ABC):
+    """Abstract class for metadata modifying implementations
+
+    NOTE: The repository module is not considered part of the python-tuf
+    stable API yet.
+
+    This class is intended to be a base class used in any metadata editing
+    application, whether it is a real repository server or a developer tool.
+
+    Implementations must implement open() and close(), and can then use the
+    edit() contextmanager to implement actual operations.
+
+    A few operations (sign, snapshot and timestamp) are already implemented
+    in this base class.
+    """
+
+    @abstractmethod
+    def open(self, role: str, init: bool = False) -> Metadata:
+        """Load a roles metadata from storage or cache, return it
+
+        If 'init', then create metadata from scratch"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def close(self, role: str, md: Metadata, sign_only: bool = False) -> None:
+        """Write roles metadata into storage
+
+        If sign_only, then just append signatures of all available keys.
+
+        If not sign_only, update expiry and version and replace signatures
+        with ones from all available keys. Keep snapshot_info and targets_infos
+        updated."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def targets_infos(self) -> Dict[str, MetaFile]:
+        """Returns the current targets version information
+
+        Not that there is a difference between this and the published snapshot
+        meta: This dictionary reflects the targets metadata that currently
+        exists in the repository, but the dictionary published by snapshot()
+        will also include metadata that no longer exists in the repository.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def snapshot_info(self) -> MetaFile:
+        """Returns the information matching current snapshot metadata"""
+        raise NotImplementedError
+
+    @contextmanager
+    def edit(
+        self, role: str, init: bool = False
+    ) -> Generator[Signed, None, None]:
```

Comment:
the core usability complaint I have with edit() is that it doesn't handle better static typing:

```python
with edit("targets") as targets:
    # mypy sees targets as type Signed, not Targets
```

Most issues would just be solved by adding a typed helper:
```python
    def edit_targets(
        self, role: str, init: bool = False
    ) -> Generator[Targets, None, None]:
```

(not sure if the other types even need a helper: timestamp and snapshot at least don't need so much editing: Repository handles the usual cases already)

Reply:
I am not very exited about adding a helper just for typing. What's the implication if mypy doesn't know the subtype?

Reply:
The implication is that more complex code in the with-block (like this [attempt at easy target removal](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/librepo/repo.py#L158)) won't have type suggestions in your editor and won't be type checked by mypy unless you do some extra work.

The workaround I use in that example makes VSCode suggestions work but doesn't make mypy happy... I think the least disruptive workaround is probably a `assert(isinstance(Targets, targets))` as the first line inside the with-block.

Reply:
but just to be clear: I'm not suggesting to add a helper in this PR -- just bringing up potential areas of improvement